daemon/ZilliqaDaemon.cpp

```diff
@@ -113,203 +175,229 @@ vector<pid_t> getProcIdByName(const string& procName, ofstream& log) {
+              string token = fullLine.substr(0, space_pos);
+              fullLine = fullLine.substr(space_pos + 1);
+
+              if (token == SYNCTYPE_OPT) {
+                space_pos = (string::npos == fullLine.find('\0'))
+                                ? fullLine.size()
+                                : fullLine.find('\0');
+                m_syncType = stoi(fullLine.substr(0, space_pos));
+                fullLine = fullLine.substr(space_pos + 1);
+                continue;
+              }
+            }
           }
         }
       }
     }
   }
 
   closedir(dp);
-
-  return result;
-}
-string execute(const string& cmd) {
-  array<char, 128> buffer{};
-  string result;
-  shared_ptr<FILE> pipe(popen(cmd.c_str(), "r"), pclose);
-  if (!pipe) throw std::runtime_error("popen() failed!");
-  while (!feof(pipe.get())) {
-    if (fgets(buffer.data(), 128, pipe.get()) != nullptr)
-      result += buffer.data();
-  }
   return result;
 }
 
-int getRestartValue(pid_t pid, const string& prgname) {
-  if (prgname == "latezilliqa" || prgname == "zilliqa") {
-    if (isDS[pid]) {
-      return DS_SYNC;
+void ZilliqaDaemon::StartNewProcess() {
+  m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+        << "Create new Zilliqa process..." << endl;
+  signal(SIGCHLD, SIG_IGN);
+
+  if (0 == fork()) {
+    bool bSuspend = false;
+
+    while (ifstream(SUSPEND_LAUNCH).good()) {
+      if (!bSuspend) {
+        m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+              << "Temporarily suspend launch new zilliqa process, please wait "
+                 "until \""
+              << SUSPEND_LAUNCH << "\" file disappeared." << endl;
+        bSuspend = true;
+      }
+
+      sleep(1);
+    }
+
+    string strSyncType;
+
+    if (m_cseed) {
+      // 1. Download Incremental DB Persistence
+      // 2. Restart zilliqa with syncType 6
+      while (!DownloadPersistenceFromS3()) {
+        m_log << ZilliqaDaemon::CurrentTimeStamp().c_str()
+              << "Downloading persistence from S3 has failed. Will try again!"
+              << endl;
+        this_thread::sleep_for(chrono::seconds(10));
+      }
+
+      strSyncType = std::to_string(NEW_LOOKUP_SYNC);
     } else {
-      return NORMAL_SYNC;
+      /// For recover-all scenario, a SUSPEND_LAUNCH file wil be created prior
+      /// to Zilliqa process being killed. Thus, we can use the variable
+      /// 'bSuspend' to distinguish syncType as RECOVERY_ALL_SYNC or NO_SYNC.
+      strSyncType =
+          bSuspend ? to_string(RECOVERY_ALL_SYNC) : to_string(m_syncType);
+      m_recovery =
+          (strSyncType == to_string(RECOVERY_ALL_SYNC)) ? 1 : m_recovery;
+      m_log << "Suspend launch is " << bSuspend
+            << ", set syncType = " << strSyncType
+            << ", recovery = " << m_recovery << endl;
     }
-  } else if (prgname == "lzilliqa") {
-    return LOOKUP_SYNC;
-  }
 
-  return -1;
+    string cmdToRun = "python " + m_curPath + launch_zilliqa_script + " " +
```

Comment:
Can we update the `launch_zilliqa.py` script to support python3 interpreter?

Reply:
Another concern is why would we still need the script `launch_zilliqa.py`?

If I'm not wrong we are adding a python wrapper around `zilliqa` binary for some additional functions that could be done in C++.

Reply:
Actually it's original design and I just keep it. But anyway, yes this script could be removed. I'll try to remove `launch_zilliqa.py` in this PR also.